cmd/notify-mysql.go

```diff
@@ -0,0 +1,257 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// MySQL Notifier implementation. A table with a specific
+// structure (column names, column types, and primary key/uniqueness
+// constraint) is used. The user may set the table name in the
+// configuration. A sample SQL command that creates a command with the
+// required structure is:
+//
+//     CREATE TABLE myminio (
+//         key_name VARCHAR(2048),
+//         value JSONB,
+//         PRIMARY KEY (key_name),
+//     );
+//
+// MySQL's "INSERT ... ON DUPLICATE ..." feature (UPSERT) is used
+// here. The implementation has been tested with MySQL Ver 14.14
+// Distrib 5.7.17.
+//
+// On each create or update object event in Minio Object storage
+// server, a row is created or updated in the table in MySQL. On
+// each object removal, the corresponding row is deleted from the
+// table.
+
+package cmd
+
+import (
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/Sirupsen/logrus"
+
+	// Add mysql driver
+	_ "github.com/go-sql-driver/mysql"
+	mysql "github.com/go-sql-driver/mysql"
+)
+
+const (
+	upsertRowMySQL = `INSERT INTO %s (key_name, value)
+VALUES (?, ?)
+ON DUPLICATE KEY UPDATE value=VALUES(value);
+`
+	deleteRowMySQL = ` DELETE FROM %s
+WHERE key_name = ?;`
+	createTableMySQL = `CREATE TABLE %s (
+    key_name VARCHAR(2048),
+    value JSON,
+    PRIMARY KEY (key_name)
+);`
+	tableExistsMySQL = `SELECT 1 FROM %s;`
+)
+
+type mySQLNotify struct {
+	Enable bool `json:"enable"`
+
+	// pass data-source-name connection string in config
+	// directly. This string is formatted according to
+	// https://github.com/go-sql-driver/mysql#dsn-data-source-name
+	DsnString string `json:"dsnString"`
+	// specifying a table name is required.
+	Table string `json:"table"`
+
+	// uses the values below if no connection string is specified
+	// - however the connection string method offers more
+	// flexibility.
+	Host     string `json:"host"`
+	Port     string `json:"port"`
+	User     string `json:"user"`
+	Password string `json:"password"`
+	Database string `json:"database"`
+}
+
+func (m *mySQLNotify) Validate() error {
+	if !m.Enable {
+		return nil
+	}
+	if m.DsnString == "" {
+		if _, err := checkNetURL(m.Host); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type mySQLConn struct {
+	dsnStr        string
+	table         string
+	preparedStmts map[string]*sql.Stmt
+	*sql.DB
+}
+
+func dialMySQL(msql mySQLNotify) (mySQLConn, error) {
+	if !msql.Enable {
+		return mySQLConn{}, errNotifyNotEnabled
+	}
+
+	// check that table is specified
+	if msql.Table == "" {
```

Comment:
Yep, done this also. But checking for empty database name is not necessary. If `dsnString` is provided, then `database` can be empty. `dsnString` provides more flexibility in connecting to the database than the remaining parameters and that is why it is retained.